src/couch/src/couch_db_updater.erl

```diff
@@ -692,6 +641,87 @@ update_local_doc_revs(Docs) ->
     end, Docs).
 
 
+purge_docs(Db, []) ->
+    {ok, Db, []};
+
+purge_docs(Db, PurgeReqs) ->
+    Ids = lists:usort(lists:map(fun({_UUID, Id, _Revs}) -> Id end, PurgeReqs)),
+    FDIs = couch_db_engine:open_docs(Db, Ids),
+    USeq = couch_db_engine:get_update_seq(Db),
+
+    IdFDIs = lists:zip(Ids, FDIs),
+    {NewIdFDIs, Replies} = apply_purge_reqs(PurgeReqs, IdFDIs, USeq, []),
+
+    Pairs = lists:flatmap(fun({DocId, OldFDI}) ->
+        {DocId, NewFDI} = lists:keyfind(DocId, 1, NewIdFDIs),
```

Comment:
It seems like this line would be executed very often. Every call is O(n). We would be traversing the list over O(2n+2*n^2) times (purge_docs and apply_purge_reqs combined) (I am sure my calculation is not precise). 
Did you consider using `gb_sets` to hold `IdFDIs`?

Reply:
Hopefully its not executed too often or that suggests we have a user that didn't understand the messaging that this feature is a last resort API and not intended to be part of normal application workflow.

Also HTTP should be limiting N here to less than 100 which means the linear searching should be cheaper than building a dict or some other data structure for lookups. There's a similar case in the implementation of `couch_util:reorder_results/1` for reference on the size issue.